File: core/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/ui/menu/defaultMenu/DefaultYouTubePlayerMenu.java
package com.pierfrancescosoffritti.androidyoutubeplayer.ui.menu.defaultMenu;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.PopupWindow;

import com.pierfrancescosoffritti.androidyoutubeplayer.R;
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.menu.MenuItem;
import com.pierfrancescosoffritti.androidyoutubeplayer.ui.menu.YouTubePlayerMenu;

import java.util.ArrayList;
import java.util.List;

public class DefaultYouTubePlayerMenu implements YouTubePlayerMenu {

    private final Context context;
    private final List<MenuItem> menuItems;

    @Nullable private PopupWindow popupWindow;

    public DefaultYouTubePlayerMenu(@NonNull Context context) {
        this.context = context;

        this.menuItems = new ArrayList<>();
    }

    @Override
    public void show(View anchorView) {
        popupWindow = createPopupWindow();
        popupWindow.showAsDropDown(anchorView, 0, - context.getResources().getDimensionPixelSize(R.dimen._8dp) * 4);

        if(menuItems.size() == 0)
            Log.e(YouTubePlayerMenu.class.getName(), "The menu is empty");
    }

    @Override
    public void dismiss() {
        if (popupWindow != null)
            popupWindow.dismiss();
    }

    @Override
    public void addItem(MenuItem menuItem) {
        menuItems.add(menuItem);
    }

    @Override
    public void removeItem(int itemIndex) {
        menuItems.remove(itemIndex);
    }

    @Override
    public int getItemCount() {
        return menuItems.size();
    }

    @NonNull
    private PopupWindow createPopupWindow() {
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        if(inflater == null)
            throw new RuntimeException("can't access LAYOUT_INFLATER_SERVICE");

        View view = inflater.inflate(R.layout.player_menu, null);

        RecyclerView recyclerView = view.findViewById(R.id.recycler_view);
        setUpRecyclerView(recyclerView);

        PopupWindow popupWindow = new PopupWindow(view, ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        popupWindow.setFocusable(true);
        popupWindow.setWidth(WindowManager.LayoutParams.WRAP_CONTENT);
        popupWindow.setHeight(WindowManager.LayoutParams.WRAP_CONTENT);
        popupWindow.setContentView(view);

        return popupWindow;
    }

    private void setUpRecyclerView(RecyclerView recyclerView) {
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));

        MenuAdapter adapter = new MenuAdapter(context, menuItems);
        recyclerView.setAdapter(adapter);
    }
}

File: core/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/core/ui/menu/defaultMenu/DefaultYouTubePlayerMenu.kt
package com.pierfrancescosoffritti.androidyoutubeplayer.core.ui.menu.defaultMenu

import android.content.Context
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.PopupWindow

import com.pierfrancescosoffritti.androidyoutubeplayer.R
import com.pierfrancescosoffritti.androidyoutubeplayer.core.ui.menu.MenuItem
import com.pierfrancescosoffritti.androidyoutubeplayer.core.ui.menu.YouTubePlayerMenu

import java.util.ArrayList

internal class DefaultYouTubePlayerMenu(private val context: Context) : YouTubePlayerMenu {
    private val menuItems = ArrayList<MenuItem>()
    private var popupWindow: PopupWindow? = null

    override val itemCount: Int
        get() = menuItems.size

    override fun show(anchorView: View) {
        popupWindow = createPopupWindow()
        popupWindow?.showAsDropDown(anchorView, -context.resources.getDimensionPixelSize(R.dimen.ayp_8dp) * 12, -context.resources.getDimensionPixelSize(R.dimen.ayp_8dp) * 12)

        if (menuItems.size == 0)
            Log.e(YouTubePlayerMenu::class.java.name, "The menu is empty")
    }

    override fun dismiss() {
        popupWindow?.dismiss()
    }

    override fun addItem(menuItem: MenuItem): YouTubePlayerMenu {
        menuItems.add(menuItem)
        return this
    }

    override fun removeItem(itemIndex: Int): YouTubePlayerMenu {
        menuItems.removeAt(itemIndex)
        return this
    }

    override fun removeItem(menuItem: MenuItem): YouTubePlayerMenu {
        menuItems.remove(menuItem)
        return this
    }

    private fun createPopupWindow(): PopupWindow {
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view = inflater.inflate(R.layout.ayp_player_menu, null)

        val recyclerView = view.findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = MenuAdapter(context, menuItems)
        recyclerView.setHasFixedSize(true)

        val popupWindow = PopupWindow(view, ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        popupWindow.contentView = view
        popupWindow.isFocusable = true
        popupWindow.width = WindowManager.LayoutParams.WRAP_CONTENT
        popupWindow.height = WindowManager.LayoutParams.WRAP_CONTENT

        return popupWindow
    }
}
